### PR TITLE
feat: HTML + Markdown output formats (closes #6, builds on #8)

### LINKED ISSUE #6: feat: add option for exporting OCR output to markdown + HTML

### PR DESCRIPTION
**Pull request by ahnafnafee**
## Summary

Closes [#6](https://github.com/ahnafnafee/local-llm-pdf-ocr/issues/6) by adding **HTML and Markdown** as alongside-PDF output formats. Builds on [@milahu](https://github.com/milahu)'s draft [PR #8](https://github.com/ahnafnafee/local-llm-pdf-ocr/pull/8) — milahu's seven commits are preserved at their original SHAs at the branch base, with `Co-authored-by` trailers on the relocated logic.

### What you can now do

```bash
# Auto-named per format
uv run local-llm-pdf-ocr scan.pdf --format html      # → scan_ocr.html
uv run local-llm-pdf-ocr scan.pdf --format md        # → scan_ocr.md

# Or pick by extension
uv run local-llm-pdf-ocr scan.pdf out.html
uv run local-llm-pdf-ocr scan.pdf notes.md

# HTML with dark-mode page inversion (opt-in)
uv run local-llm-pdf-ocr scan.pdf --format html --html-invert-dark

# Server: same, via the /process form field
curl -F file=@scan.pdf -F client_id=x -F format=html http://localhost:8000/process
```

The web UI (`server.py`) gets a small format dropdown beneath the upload zone.

### Architecture
- **`src/pdf_ocr/core/html.py` (new)** — `HTMLHandler` with PDF *and* image input support (PR #8 only handled images), in-memory `StringIO` build, base64-inlined page images so the file is self-contained. Supports three sizing modes (`scaled` default, `letter-spacing`, `full-height`) and an opt-in `invert_dark` flag that injects `filter: invert() hue-rotate(180deg)` under `prefers-color-scheme: dark`.
- **`src/pdf_ocr/core/markdown.py` (new)** — `MarkdownHandler`: `# OCR output` header, `## Page N` per page, one block per non-empty box in reading order. No paragraph-break heuristic (multi-column layouts break gap-based heuristics).
- **`src/pdf_ocr/core/_layout.py` (new)** — extracted `is_full_page_fallback` and `split_multi_line_bbox` helpers shared by every writer. Both PDF and HTML now route the aligner's `[0,0,1,1]+\n` fallback bbox identically, and split `\n`-joined visual lines into per-line sub-spans.
- **`src/pdf_ocr/output.py` (new)** — `resolve_output_writer`, `media_type_for`, `suffix_for_format`, `format_from_path`, `SUPPORTED_FORMATS`. Single source of truth for format dispatch; CLI and server both consume it.
- **`src/pdf_ocr/cli.py`** — adds `--format {pdf,html,md}` with extension-wins precedence, `--html-invert-dark` opt-in flag; passes the dispatched writer to `OCRPipeline(output_writer=…)`. Also reverts PR #8's `raise; sys.exit(1)` regression in `main()`.
- **`src/pdf_ocr/server.py`** — adds `format: str = Form("pdf")` to `/process`; wires Content-Type and download filename through `media_type_for`/`suffix_for_format`. Validates format and returns 400 on unsupported values.

### `--html-invert-dark` flag (addresses PR feedback)

Added in response to @milahu's [feedback](https://github.com/ahnafnafee/local-llm-pdf-ocr/pull/10#discussion_r3253296195) about dark-mode page inversion. The CSS `filter: invert() hue-rotate(180deg)` on `div.page` is useful for reading scanned white-background documents at night, but is now **opt-in** since most PDF readers and browsers do not invert scanned images by default.

- `HTMLHandler(invert_dark=True)` injects an additional `@media (prefers-color-scheme: dark)` block
- CLI: `--html-invert-dark`
- `resolve_output_writer(html_invert_dark=True)` passes it through
- Example output: `examples/output_digital_invert_dark.html`

### Alignment quality — `scaled` default

The HTML invisible-text overlay defaults to `scaled` mode — font shrinks to fit both bbox dimensions, staying legible at any zoom level. Two alternative modes (`letter-spacing`, `full-height`) are available via `HTMLHandler(mode=…)` / `--html-mode`.

### PR #8 cleanup

| Issue | Resolution |
|---|---|
| `_embed_structured_text_html` raised `NotImplementedError` for PDF inputs | Fixed — `HTMLHandler` rasterizes per page like `PDFHandler` |
| Stray `print(f"output_ext: ...")` debug statement | Removed |
| Multiple `# TODO remove` blocks of unreachable code | Removed |
| `raise; sys.exit(1)` in cli.py | Reverted |
| HTML logic embedded in `PDFHandler` | Relocated to `core/html.py` per author's recommendation |

### Tests

113 new tests, all on the fast tier (no LLM, no Surya):

- `test_layout.py` — 14 tests: full-page-fallback detection, multi-line bbox split
- `test_html_handler.py` — 15 tests: doctype + title, base64 inline, all sizing modes, full-page fallback, multi-line split, edge cases
- `test_markdown_handler.py` — 7 tests: structure, page order, reading order, empty-box skip
- `test_output_dispatch.py` — 28 tests: extension inference, suffix/media-type lookup, writer resolution
- `test_cli.py` — 9 tests: `resolve_output_path` precedence, `--format` parser
- `test_server.py` — 8 tests: format dispatch, download naming, error handling

Total suite: **294 tests passing** (271 fast + 23 slow).

### Documentation

- README: `--html-invert-dark` in the CLI options table and examples section
- CLAUDE.md: updated commands, HTMLHandler docstring with `invert_dark` parameter

## Authorship preservation

milahu's seven commits ([`f775df5`](https://github.com/ahnafnafee/local-llm-pdf-ocr/commit/f775df5)…[`10078d8`](https://github.com/ahnafnafee/local-llm-pdf-ocr/commit/10078d8)) remain at their original SHAs at the branch base — no rebase, squash, amend, or force-push. Cleanup commits on top use `Co-authored-by: Milan Huth <milahu@milahu.duckdns.org>` wherever the new code derives from theirs.

## Test plan

- [x] All 294 tests pass (`uv run pytest`)
- [x] CLI `--help` shows `--html-invert-dark`
- [x] Without flag: no `filter: invert()` in HTML output
- [x] With `--html-invert-dark`: `filter: invert() hue-rotate(180deg)` present under `@media (prefers-color-scheme: dark)`
- [x] All example outputs regenerated against current code

### CODE REVIEW

**Comment by ahnafnafee** [comment]
@milahu Do the latest changes satisfy your expected output?

File: src/pdf_ocr/cli.py [hunk 1]
```diff
@@ -83,6 +87,27 @@ def build_parser() -> argparse.ArgumentParser:
...
+        help="Span sizing strategy for HTML output (ignored for pdf/md). "
+             "letter-spacing (default): font sized to bbox height, chars "
+             "spread to fill bbox width — best when Surya bboxes match "
+             "visible text width. full-height: font = bbox height, no "
+             "letter-spacing — text uses natural monospace width and may "
+             "overflow the bbox right edge. scaled: shrinks font so text "
+             "fits both bbox dimensions — most compact, but smaller "
+             "selection extents. Try the alternatives if letter-spacing "
+             "produces visible overlay characters past the underlying text.",
+    )
```
Comment (milahu) [inline src/pdf_ocr/cli.py:120]: the default should be `html_mode="scaled"`
`html_mode="letter-spacing"` can make the text hard to read
when letters overlap with a negative letter-spacing

with `html_mode="scaled"`
at least i can zoom in to read the text

File: src/pdf_ocr/core/html.py [hunk 2]
```diff
@@ -0,0 +1,277 @@
...
+  color: transparent;
+  white-space: nowrap;
+  font-family: monospace;
+  line-height: 1;
+  /* Keep invisibility intact even when the browser is in dark mode —
+     a previous draft applied `filter: invert()` to the page div, which
+     interacts badly with `color: transparent` in Chromium and renders
+     the glyph outlines as a faint inverted color. The OCR HTML preserves
+     the source page's appearance regardless of OS theme; users who want
+     dark theming for documents can use a browser extension. */
```
Comment (milahu) [inline src/pdf_ocr/core/html.py]: this comment is not helpful in the HTML output
it should be only in the python source

File: src/pdf_ocr/core/html.py [hunk 3]
```diff
@@ -0,0 +1,277 @@
...
+        return out.getvalue()
+
+    def _render_page(self, out, page_num, img_bytes, width, height, items):
+        b64 = base64.b64encode(img_bytes).decode("ascii")
+        data_url = f"data:image/jpeg;base64,{b64}"
+        out.write(
+            f'<div class="page" data-page="{page_num + 1}" '
+            f'style="width:{_num(width)}px;height:{_num(height)}px;'
+            f"background-image:url('{data_url}')\">\n"
+        )
```
Comment (milahu) [inline src/pdf_ocr/core/html.py]: by default, it should use external image files from HTML
and there should be a CLI option
to enable embedding base64-encoded images into HTML
something like `--html-embed-images` or `--html-inline-images`

why?

- base64 encoding increases the file size by 35%, which is not wanted in most cases
- the image files exist as input files as one image per page
  - exception: when the input is a PDF file (or TIFF file) with multiple pages, then it is not possible to use the input file as image source in HTML. then each page image should be stored side by side with the output HTML file, only with a different file extension, for example `some-page.html` and `some-page.jpg`

Comment (ahnafnafee) [reply src/pdf_ocr/core/html.py]: External images should be default now

File: src/pdf_ocr/core/html.py [hunk 4]
```diff
@@ -76,26 +73,60 @@
...
+<script>
+(function(){
+  var px = window.innerWidth * (window.devicePixelRatio || 1);
+  document.querySelectorAll('div.page').forEach(function(p){
+    var w = parseFloat(p.style.getPropertyValue('--page-w'));
+    if (!isFinite(w) || w <= 0) return;
+    var s = Math.min(1, px / w);
+    if (s < 1) p.style.width = (w * s) + 'px';
+  });
+})();
```
Comment (milahu) [inline src/pdf_ocr/core/html.py:110]: this fails for `window.devicePixelRatio != 1`
(when then display manager zoom level is not 100%)

`window.innerWidth = document.body.clientWidth + scrollbarWidth`

`clientWidth - 1` to account for rounding errors
otherwise i still see a horizontal scrollbar
(my desktop zoom is 150%)

simplified function

```js
(function(){
  var px = document.body.clientWidth - 1;
  for (const p of document.querySelectorAll('div.page')) {
    var w = parseFloat(p.style.getPropertyValue('--page-w'));
    if (!isFinite(w) || w <= 0 || w <= px) continue;
    p.style.width = px + 'px';
  }  
})();
```

File: src/pdf_ocr/core/html.py [hunk 5]
```diff
@@ -46,25 +25,43 @@
...
-# we don't control. Treated as a tunable class-level constant.
+# Browser monospace stacks (Courier-family) sit near 0.6 width:height.
 _MONOSPACE_FONT_ASPECT = 0.6
 
 _JPEG_QUALITY = 80
 
+# No `filter: invert()` on the page div: it interacts with
+# `color: transparent` in Chromium and renders glyph outlines as a faint
+# inverted color. The output preserves the source page's appearance; OS
+# dark theming is a browser-extension concern.
```
Comment (milahu) [inline src/pdf_ocr/core/html.py]: darkreader does not invert images...

in my code i have this style
which just works, no "glyph outlines"

```css
@media (prefers-color-scheme: dark) {
  div.page {
    filter: invert() hue-rotate(180deg);
  }
}
```

the `hue-rotate(180deg)` is needed
to convert darkblue to lightblue (instead of yellow)

"no darkmode" is also a stupid limitation of chrome's PDF reader
which makes it impossible for me to read PDF documents at night

File: src/pdf_ocr/core/html.py [hunk 6]
```diff
@@ -0,0 +1,393 @@
...
+        out_parent = Path(output_path).resolve().parent
+        out_stem = Path(output_path).stem
+        direct = self._direct_reference_url(input_path, out_parent)
+        if direct is not None:
+            url, width, height = direct
+            yield 0, url, float(width), float(height)
+            return
+
+        for page_num, img_bytes, w, h in self._rasterize_pages(input_path, dpi):
+            sidecar_name = f"{out_stem}_p{page_num + 1}.jpg"
```
Comment (milahu) [inline src/pdf_ocr/core/html.py]: for single-page input files, there should be no `_p1` suffix

so...

```py
        rasterized_pages = self._rasterize_pages(input_path, dpi)
        for page_num, img_bytes, w, h in rasterized_pages:
            if len(rasterized_pages) == 1:
                sidecar_name = f"{out_stem}.jpg"
            else:
                sidecar_name = f"{out_stem}_p{page_num + 1}.jpg"
```

File: src/pdf_ocr/core/html.py [hunk 7]
```diff
@@ -0,0 +1,393 @@
...
+        out.write('<meta charset="utf-8" />\n')
+        out.write(f"<title>{_html.escape(title)}</title>\n")
+        out.write("<style>\n")
+        out.write(_PAGE_CSS)
+        out.write("</style>\n</head>\n<body>\n")
+        for page_num, image_url, width, height in pages:
+            self._render_page(
+                out, page_num, image_url, width, height,
+                pages_data.get(page_num, []),
+            )
```
Comment (milahu) [inline src/pdf_ocr/core/html.py]: have you tested this with multi-page input files?
are the pages rendered properly, like in a PDF reader?

if one page has less width than other pages
it should be centered horizontally with

```css
body {
  text-align: center;
}
```

or

```css
div.page {
  margin: auto;
}
```

Comment (ahnafnafee) [reply src/pdf_ocr/core/html.py]: I will test this out and report back

File: src/pdf_ocr/core/html.py (hunk 5 above)
Comment (ahnafnafee) [reply src/pdf_ocr/core/html.py]: I think this could be passed in as a CLI optional parameter instead of being the default

Comment (milahu) [reply src/pdf_ocr/core/html.py]: this is a special case, because we have images of text

the darkreader extension does not invert images, because darkreader does not know the difference between "images of text" and "images of cats", because there is no magic style class like `img.imageoftext.lightmode`, so i need the stylebot extension to invert images of text...

with the principle of least surprise, i would expect that when i set my display manager's color theme to dark (`@media (prefers-color-scheme: dark)`) then that should also invert images of text

in short, this should be default on, with an option to turn it off

> no magic style class like `img.imageoftext.lightmode`

some book authors follow the bad taste of having some pages inverted (white text on black background) ([example with L=0.02](https://milahu.github.io/the-preparation-by-doug-casey-2025-epub/OEBPS/images/008.avif)), so we could [compute the lightness](https://github.com/milahu/hocr-files-template-repo/blob/5e26f5b601970b8a4188e05c8e48ae1c9849f598/070-deskew.py#L56) of each page ([example output](https://github.com/milahu/the-preparation-by-doug-casey-2025/blob/main/070-deskew.lightness.txt)), to ensure that in our dark mode, the originally dark pages are not inverted, so all pages are dark. books can also have pages with an average lightness near 50% (gray) ([example with L=0.35](https://milahu.github.io/the-preparation-by-doug-casey-2025-epub/OEBPS/images/025.avif)), so pages should be inverted only above some threshold, maybe 0.7 by default

Comment (ahnafnafee) [reply src/pdf_ocr/core/html.py]: Most tools (PDF readers, browsers) do not invert scanned page images by default, so automatically inverting them could be surprising to many users. I believe allowing it to be opt-in CLI flag would allow more customization for the end user.

Comment (ahnafnafee) [reply src/pdf_ocr/core/html.py]: I've added it as an opt-in CLI flag in the latest push (`1d13516`).

**Usage:**

```bash
uv run local-llm-pdf-ocr input.pdf --format html --html-invert-dark
```

This adds the following CSS block to the generated HTML, which activates only when the browser/OS is in dark mode:

```css
@media (prefers-color-scheme: dark) {
  div.page {
    filter: invert() hue-rotate(180deg);
  }
}
```

Without `--html-invert-dark`, the page image renders as-is in all colour schemes (just the body background and outline adjust for dark mode). This keeps the default behaviour unsurprising for users with scanned colour documents or photos, while making it easy to opt in for the white-background-at-night use case you described.

I also generated an example with the flag applied: [`examples/output_digital_invert_dark.html`](https://github.com/ahnafnafee/local-llm-pdf-ocr/blob/fix/issue-6-html-md-export/examples/output_digital_invert_dark.html) — you can download it and test in your browser with dark mode enabled. The `hue-rotate(180deg)` handles the dark-blue → light-blue remapping you mentioned.

File: src/pdf_ocr/core/html.py [hunk 8]
```diff
@@ -0,0 +1,416 @@
...
+            url, width, height = direct
+            yield 0, url, float(width), float(height)
+            return
+
+        rasterized = list(self._rasterize_pages(input_path, dpi))
+        for page_num, img_bytes, w, h in rasterized:
+            if len(rasterized) == 1:
+                sidecar_name = f"{out_stem}.jpg"
+            else:
+                sidecar_name = f"{out_stem}_p{page_num + 1}.jpg"
```
Comment (milahu) [inline src/pdf_ocr/core/html.py]: nitpick: `page_num + 1` should be zero-padded
so the image files are always sorted correctly

```py
        rasterized = list(self._rasterize_pages(input_path, dpi))
        pnw = len(str(len(rasterized))) # page number width
        for page_num, img_bytes, w, h in rasterized:
            if len(rasterized) == 1:
                sidecar_name = f"{out_stem}.jpg"
            else:
                sidecar_name = f"{out_stem}_p{str(page_num + 1).zfill(pnw)}.jpg"
```

nitpick:
`page_num` should be renamed to `page_idx`
(1-based versus 0-based)

---

out of scope:
ideally we should avoid
`list(self._rasterize_pages(input_path, dpi))`
currently we need that only for `len(rasterized)`
but that forces to buffer the whole input in memory
which fails on large inputs or low memory...
ideally, the whole pipeline should be streaming

related: streaming PDF writer

- https://github.com/pymupdf/PyMuPDF/issues/4968

File: examples/output_digital.html [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+<span class="line" style="left:12.0101%;top:80.7617%;width:79.1403%;font-size:1.3059cqw;height:1.3672%;">4. If there is a CS-affiliated faculty who will co-supervise the CS 701 project, their name should be</span>
+<span class="line" style="left:15.1707%;top:82.6172%;width:75.9798%;font-size:1.3616cqw;height:1.4648%;">indicated next to the "CS faculty member name" field above, and they should co-sign the form.</span>
+<span class="line" style="left:15.1707%;top:84.4727%;width:69.9115%;font-size:1.3092cqw;height:1.2695%;">At the registrar level, the CS 701 registration will be made only under CS faculty names.</span>
+<span class="line" style="left:12.1365%;top:86.2305%;width:74.5891%;font-size:1.4978cqw;height:1.4648%;">5. Submit the completed form to both csphd@gmu.edu and aydin@gmu.edu by e-mail (put</span>
+<span class="line" style="left:15.1707%;top:88.0859%;width:46.9027%;font-size:1.3714cqw;height:1.4648%;">"[CS701 form]" in the subject line) by November 21, 2025.</span>
+</div>
+<script>
+(function(){
+  var px = document.body.clientWidth - 1;
```
Comment (milahu) [inline examples/output_digital.html:73]: all `examples/*.html` still have the old version

```
$ rg -F 'var px = window.innerWidth' examples/ | wc -l
5

$ rg -F 'var px = document.body.clientWidth' examples/ | wc -l
0
```
